src/reflect/scala/reflect/internal/transform/Erasure.scala

```diff
@@ -112,8 +112,9 @@ trait Erasure {
     protected def eraseDerivedValueClassRef(tref: TypeRef): Type = erasedValueClassArg(tref)
 
     def apply(tp: Type): Type = tp match {
-      case ConstantType(_) =>
-        tp
+      case ConstantType(ct) =>
+        if (ct.tag == ClazzTag) ConstantType(Constant(apply(ct.typeValue)))
+        else tp
```

Comment:
Yes, in Scala/JVM you can only see it in the AST. Because of constant-folding for `final val`s coming from Scala-defined classes, the `final val` is not actually accessed at run-time. You can fully reproduce the bug by implemented an inherited `def` with the `final val`:
```scala
Welcome to Scala 2.12.0-M4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_60).
Type in expressions for evaluation. Or try :help.

scala> trait Foo { def C: Class[_] }
defined trait Foo

scala> object Bar extends Foo { final val C = classOf[Unit] }
defined object Bar

scala> Bar.C // constant-folded, works
res0: Class[Unit] = void

scala> val foo: Foo = Bar
foo: Foo = Bar$@16cb0050

scala> foo.C // not constant-folded, does not work
res1: Class[_] = class scala.runtime.BoxedUnit
```